File: examples/sma-crossover.js
export const script = {
  name: "sma-crossover",
  version: "1",
  sources: ["candles", "orderbook"],
  clock: "candles",
  lookback: 9,
  params: {
    entry_amount_dollar: { type: "number", required: false, default: 100 },
    max_spread: { type: "number", required: true },
    max_slippage: { type: "number", required: true },
    leverage: { type: "number", required: false, default: 10 }
  }
}

export function onData(ctx, input, history) {
  if (input.source !== "candles@binancef@mmt") return

  const candles = history.source("candles@binancef@mmt")
  const book = history.source("orderbook@bulk", 0)
  if (candles.length < 9 || !book) return

  const fast = ctx.study.sma(candles, { window: 3 })
  const slow = ctx.study.sma(candles, { window: 8 })
  const long = fast.previous <= slow.previous && fast.latest > slow.latest
  const short = fast.previous >= slow.previous && fast.latest < slow.latest

  const side = long ? "long" : short ? "short" : null
  if (!side) return

  const spread = ctx.study.spread(book)
  const slippage = ctx.study.slippage(book, {
    side: long ? "buy" : "sell",
    notional: ctx.params.entry_amount_dollar
  })

  if (
    spread.spread_bps < 0 ||
    spread.spread_bps > ctx.params.max_spread ||
    slippage.slippage_bps > ctx.params.max_slippage
  ) return

  const candle = history.source("candles@binancef@mmt", 0)
  ctx.trade({
    key: `sma-cross-${candle.t}`,
    side,
    notional: ctx.params.entry_amount_dollar,
    order: { type: "market" },
    leverage: ctx.params.leverage
  })
}

File: examples/sma-crossover.py
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd


SOURCE = "btc@candles@binancef@mmt"

script = {
    "name": "python-sma-crossover",
    "version": "2",
    "sources": ["candles"],
    "lookback": 100,
    "params": {
        "fast_period": {"type": "number", "default": 5},
        "slow_period": {"type": "number", "default": 20},
        "margin": {"type": "number", "default": 100},
    },
}


def on_data(ctx, event, history):
    if event["source"] != SOURCE:
        return

    candles = history.source(SOURCE)
    slow_period = int(ctx.params["slow_period"])
    if len(candles) < slow_period + 1:
        return

    closes = pd.Series([candle["c"] for candle in candles])
    fast = closes.rolling(int(ctx.params["fast_period"])).mean()
    slow = closes.rolling(slow_period).mean()
    previous = np.sign(fast.iloc[-2] - slow.iloc[-2])
    current = np.sign(fast.iloc[-1] - slow.iloc[-1])

    positions = event["positions"]["open"]
    position = positions[0] if positions else None
    timestamp = candles[-1]["t"]

    if previous <= 0 and current > 0 and position is None:
        ctx.trade(
            {
                "key": f"sma-buy-{timestamp}",
                "symbol": "BTC",
                "position": "open-long",
                "margin": ctx.params["margin"],
                "order": {"type": "market"},
            }
        )
    elif previous >= 0 and current < 0 and position:
        ctx.trade(
            {
                "key": f"sma-sell-{timestamp}",
                "symbol": "BTC",
                "position": "close-long",
                "order": {"type": "market"},
            }
        )

    return {"metrics": {"fast_sma": fast.iloc[-1], "slow_sma": slow.iloc[-1]}}


def on_finish(ctx, history):
    candles = pd.DataFrame(history.source(SOURCE))
    if candles.empty:
        return

    candles["fast_sma"] = candles["c"].rolling(int(ctx.params["fast_period"])).mean()
    candles["slow_sma"] = candles["c"].rolling(int(ctx.params["slow_period"])).mean()

    candles.plot(x="t", y=["c", "fast_sma", "slow_sma"])
    plt.savefig(ctx.artifact_path("sma-crossover.png"))
    plt.close()
